packages/baseui-codemods/src/v9-theme-scale.test.js

```diff
@@ -0,0 +1,256 @@
+/*
+Copyright (c) 2018-2019 Uber Technologies, Inc.
+
+This source code is licensed under the MIT license found in the
+LICENSE file in the root directory of this source tree.
+*/
+
+// @flow
+
+import codemod from './v9-theme-scale.js';
+import {Fixture} from './test-utilities.js';
+
+async function transform(content) {
+  const fixture = new Fixture();
+  await fixture.write(content);
+  await codemod({dir: fixture.dir});
+  const transformed = await fixture.read();
+  await fixture.remove();
+  return transformed;
+}
+
+describe('shift theme usage for colors and fonts', () => {
+  describe('does update', () => {
+    it('font values in Blocks', async () => {
+      const content = `const Foo = () => (
+    <React.Fragment>
+      <Block font="font1100" />
+      <Block font="font1000" />
+      <Block font="font900" />
+      <Block font="font800" />
+      <Block font="font700" />
+      <Block font="font600" />
+      <Block font="font500" />
+      <Block font="font470" />
+      <Block font="font460" />
+      <Block font="font450" />
+      <Block font="font400" />
+      <Block font="font350" />
+      <Block font="font300" />
+      <Block font="font250" />
+      <Block font="font200" />
+      <Block font="font100" />
+    </React.Fragment>
+  );`;
+
+      expect(await transform(content)).toMatchInlineSnapshot(`
+        "const Foo = () => (
+            <React.Fragment>
+              <Block font=\\"font1450\\" />
+              <Block font=\\"font1050\\" />
+              <Block font=\\"font950\\" />
+              <Block font=\\"font850\\" />
+              <Block font=\\"font750\\" />
+              <Block font=\\"font650\\" />
+              <Block font=\\"font550\\" />
+              <Block font=\\"font450\\" />
+              <Block font=\\"font400\\" />
+              <Block font=\\"font350\\" />
+              <Block font=\\"font300\\" />
+              <Block font=\\"font250\\" />
+              <Block font=\\"font200\\" />
+              <Block font=\\"font150\\" />
+              <Block font=\\"font100\\" />
+              <Block font=\\"font100\\" />
+            </React.Fragment>
+          );"
+      `);
+    });
+
+    it('direct usage of font values', async () => {
+      const content = `const Foo = styled('div', props => ({
```

Comment:
it may be unnecessary, since I'm sure you've tested this across many sample applications. But I worry about how the codemod does not select for the context in which to perform these changes. For example, only Block and related typography components should have their `font` props shifted. And similarly, only baseui's styletron wrappers have access to the theme we want to change. Some limitations/complexity of this approach, would be defining style definitions outside of where we expect or even worse importing from another file(s). Just interested to know if you've weighed these tradeoffs between more code vs the possibility or code-modding false positive matches

Reply:
this is a very valid concern. i went back and forth on how much to include. here is where i am now: i worry more about _missed_ updates (false negatives), so i made the rules for triggering an update a bit lax.

if i update something that should not be updated, you might catch that when looking over your diff. but if i don't update something that _should_ be updated, there is no easy way to catch that other than a type error or visual regression test (manual or auto) flagging sizing changes.

what really sold me was running the mod on actual repos. i found they worked quite well in practice. in the projects i sampled, devs were not really creating or importing variables with names like `font100`. i didn't encounter any false positives but 4 or 5 repos may be too small a sample to have a lot of confidence.

if there are ways to cut down potential false positives without introducing false negatives (or ~99% still pass) that would be a clear win. for example, i considered only running these changes on files that imported baseui, only updating Blocks and typography components, etc. I'd be open to adding things like that in again if they keep that negative/positive ratio.

Reply:
yeah i agree that not-found instances are much worse. this should be fine, especially given what you wrote in the blog post